arci/src/clients/joint_position_limiter.rs

```diff
@@ -0,0 +1,362 @@
+                let joint = &joints[i];
+                let limit = if JointType::Continuous == joint.joint_type {
+                    // If limit is not specified, urdf-rs assigns f64::default.
+                    -f64::consts::PI..=f64::consts::PI
+                } else {
+                    joint.limit.lower..=joint.limit.upper
+                };
+                limits.push(limit);
+            } else {
+                return Err(Error::NoJoint(joint_name.into()));
+            }
+        }
+
+        Ok(Self {
+            client,
+            limits,
+            strategy,
+        })
+    }
+
+    pub fn set_strategy(&mut self, strategy: JointPositionLimiterStrategy) {
+        self.strategy = strategy;
+    }
+
+    fn check_joint_position(&self, positions: &mut Vec<f64>) -> Result<(), Error> {
+        assert_eq!(positions.len(), self.limits.len());
+        for (i, (limit, position)) in self.limits.iter().zip(positions).enumerate() {
+            if limit.contains(&position) {
+                continue;
+            }
+            match self.strategy {
+                JointPositionLimiterStrategy::Error => {
+                    return Err(Error::OutOfLimit {
+                        name: self.client.joint_names()[i].clone(),
+                        position: *position,
+                        limit: limit.clone(),
+                    });
+                }
+                JointPositionLimiterStrategy::Round => {
+                    debug!(
+                        "Out of limit: joint={}, position={}, limit={:?}",
+                        self.client.joint_names()[i],
+                        position,
+                        limit,
+                    );
+                    *position = position.clamp(*limit.start(), *limit.end());
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<C> JointTrajectoryClient for JointPositionLimiter<C>
+where
+    C: JointTrajectoryClient,
+{
+    fn joint_names(&self) -> &[String] {
+        self.client.joint_names()
+    }
+
+    fn current_joint_positions(&self) -> Result<Vec<f64>, Error> {
+        self.client.current_joint_positions()
+    }
+
+    fn send_joint_positions(
+        &self,
+        mut positions: Vec<f64>,
+        duration: std::time::Duration,
+    ) -> Result<WaitFuture, Error> {
+        self.check_joint_position(&mut positions)?;
+        self.client.send_joint_positions(positions, duration)
+    }
+
+    fn send_joint_trajectory(
+        &self,
+        mut trajectory: Vec<TrajectoryPoint>,
+    ) -> Result<WaitFuture, Error> {
+        for tp in &mut trajectory {
+            self.check_joint_position(&mut tp.positions)?;
+        }
+        self.client.send_joint_trajectory(trajectory)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum JointPositionLimiterStrategy {
+    /// If the position is out of the limit, handle it as the same value as the limit.
+    Round,
```

Comment:
I don't know about English but we use `clamp` for this.

https://docs.rs/k/0.24.0/k/struct.Chain.html#method.set_joint_positions_clamped

Reply:
done in d01a272